File: ui.py
def display_menu():
    print("\nBlackjack Basic Strategy Trainer")
    print("1. Quick Practice (random)")
    print("2. Learn by Dealer Strength")
    print("3. Focus on Hand Types")
    print("4. Absolutes Drill")
    print("5. View Statistics")
    print("6. Quit")
    
    while True:
        try:
            choice = int(input("\nEnter your choice (1-6): "))
            if 1 <= choice <= 6:
                return choice
            else:
                print("Please enter a number between 1 and 6.")
        except ValueError:
            print("Please enter a valid number.")

def display_hand(player_cards, dealer_card, hand_type, player_total):
    print(f"\nDealer shows: {card_name(dealer_card)}")
    
    if hand_type == 'pair':
        card_val = player_cards[0]
        print(f"Your hand: {card_name(card_val)}, {card_name(card_val)} (Pair of {card_name(card_val)}s)")
    else:
        card_str = ", ".join([card_name(card) for card in player_cards])
        total_type = "Soft" if hand_type == 'soft' else "Hard"
        print(f"Your hand: {card_str} ({total_type} {player_total})")

def card_name(value):
    if value == 11:
        return "A"
    elif value == 10:
        return "10"
    else:
        return str(value)

def get_user_action():
    print("\nWhat's your move?")
    print("(H)it, (S)tand, (D)ouble, s(P)lit")
    
    while True:
        action = input("Enter your choice: ").upper().strip()
        if action in ['H', 'S', 'D', 'P']:
            return action
        else:
            print("Please enter H, S, D, or P.")

def action_name(action):
    names = {'H': 'HIT', 'S': 'STAND', 'D': 'DOUBLE', 'P': 'SPLIT', 'Y': 'SPLIT', 'N': 'NO SPLIT'}
    return names.get(action, action)

def display_feedback(correct, user_action, correct_action, explanation):
    if correct:
        print("✓ Correct!")
    else:
        print("❌ Incorrect!")
        print(f"\nCorrect answer: {action_name(correct_action)}")
        print(f"Your answer: {action_name(user_action)}")
        print(f"\nExplanation: {explanation}")
    
    response = input("\nPress Enter to continue, or q to quit: ").strip().lower()
    if response in ['q', 'quit']:
        return 'quit'
    return 'continue'

def display_session_header(mode):
    mode_names = {
        'random': 'Quick Practice',
        'dealer_groups': 'Dealer Strength Practice',
        'hand_types': 'Hand Type Focus',
        'absolutes': 'Absolutes Drill'
    }
    
    print(f"\n{'='*50}")
    print(f"Starting {mode_names.get(mode, mode)} Session")
    print("Type 'quit' at any time to return to main menu")
    print(f"{'='*50}")

def get_user_choice(prompt, choices):
    while True:
        try:
            choice = input(prompt)
            if choice.lower() == 'quit':
                return None
            choice = int(choice)
            if choice in choices:
                return choice
            else:
                print(f"Please choose from: {', '.join(map(str, choices))}")
        except ValueError:
            print("Please enter a valid number or 'quit'.")

def display_dealer_groups():
    print("\nChoose dealer strength group:")
    print("1. Weak dealers (4, 5, 6) - Bust cards")
    print("2. Medium dealers (2, 3, 7, 8)")
    print("3. Strong dealers (9, 10, A)")
    return get_user_choice("Enter choice (1-3): ", [1, 2, 3])

def display_hand_types():
    print("\nChoose hand type to practice:")
    print("1. Hard totals only")
    print("2. Soft totals only") 
    print("3. Pairs only")
    return get_user_choice("Enter choice (1-3): ", [1, 2, 3])

def clear_screen():
    import os
    os.system('clear')

File: go/internal/ui/ui.go
// Package ui provides terminal user interface utilities for the blackjack trainer.
//
// This package handles all terminal input/output operations including:
// - Menu display and user choice collection
// - Hand and scenario display
// - User action input with validation
// - Feedback display with explanations
// - Session headers and progress indicators
package ui

import (
	"blackjack_trainer/internal/strategy"
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// DisplayMenu displays the main menu and gets user choice.
func DisplayMenu() (int, bool) {
	fmt.Println("\nBlackjack Basic Strategy Trainer")
	fmt.Println("1. Quick Practice (random)")
	fmt.Println("2. Learn by Dealer Strength")
	fmt.Println("3. Focus on Hand Types")
	fmt.Println("4. Absolutes Drill")
	fmt.Println("5. View Statistics")
	fmt.Println("6. Quit")
	fmt.Print("\nChoice (1-6): ")

	reader := bufio.NewReader(os.Stdin)
	input, err := reader.ReadString('\n')
	if err != nil {
		return 0, false
	}

	choice, err := strconv.Atoi(strings.TrimSpace(input))
	if err != nil || choice < 1 || choice > 6 {
		return 0, false
	}

	return choice, true
}

// DisplaySessionHeader displays session header with mode name.
func DisplaySessionHeader(modeName string) {
	fmt.Println("\n" + strings.Repeat("=", 40))
	fmt.Printf("Training Mode: %s\n", modeName)
	fmt.Println(strings.Repeat("=", 40))
	fmt.Println("(Press 'q' + Enter to quit at any time)")
}

// DisplayHand displays the current hand and dealer card.
func DisplayHand(playerCards []int, dealerCard int, handType string, playerTotal int) {
	fmt.Printf("\nDealer shows: %s\n", strategy.CardToString(dealerCard))

	fmt.Print("Your hand: ")
	for i, card := range playerCards {
		if i > 0 {
			fmt.Print(", ")
		}
		fmt.Print(strategy.CardToString(card))
	}

	handDesc := strings.Title(handType)
	fmt.Printf(" (%s %d)\n", handDesc, playerTotal)
}

// GetUserAction gets user's action choice.
func GetUserAction() (rune, bool) {
	fmt.Println("\nWhat's your move?")
	fmt.Print("(H)it, (S)tand, (D)ouble, s(P)lit: ")

	reader := bufio.NewReader(os.Stdin)
	input, err := reader.ReadString('\n')
	if err != nil {
		return 0, true
	}

	input = strings.TrimSpace(input)
	if len(input) == 0 {
		return 0, true
	}

	action := rune(strings.ToUpper(input)[0])

	// Check for quit
	if action == 'Q' {
		return 0, true
	}

	return action, false
}

// DisplayFeedback displays feedback after user's answer.
// Returns true if user wants to quit.
func DisplayFeedback(correct bool, userAction, correctAction rune, explanation string) bool {
	if correct {
		fmt.Println("\n✓ Correct!")
	} else {
		fmt.Println("\n❌ Incorrect!")
		fmt.Printf("\nCorrect answer: %s\n", strategy.ActionToString(correctAction))
		fmt.Printf("Your answer: %s\n", strategy.ActionToString(userAction))
		fmt.Printf("\nPattern: %s\n", explanation)
	}

	fmt.Print("\nPress Enter to continue (or 'q' + Enter to quit): ")

	reader := bufio.NewReader(os.Stdin)
	input, err := reader.ReadString('\n')
	if err != nil {
		return false
	}

	input = strings.TrimSpace(input)
	return len(input) > 0 && strings.ToUpper(input)[0] == 'Q'
}

// DisplayDealerGroups displays dealer groups menu and gets user choice.
func DisplayDealerGroups() (int, bool) {
	fmt.Println("\nChoose dealer strength group to practice:")
	fmt.Println("1. Weak cards (4, 5, 6) - 'Bust cards'")
	fmt.Println("2. Medium cards (2, 3, 7, 8)")
	fmt.Println("3. Strong cards (9, 10, A)")
	fmt.Println("0. Cancel")
	fmt.Print("\nChoice (0-3): ")

	reader := bufio.NewReader(os.Stdin)
	input, err := reader.ReadString('\n')
	if err != nil {
		return 0, false
	}

	choice, err := strconv.Atoi(strings.TrimSpace(input))
	if err != nil {
		return 0, false
	}

	if choice == 0 {
		return 0, false
	}

	if choice < 1 || choice > 3 {
		return 0, false
	}

	return choice, true
}

// DisplayHandTypes displays hand types menu and gets user choice.
func DisplayHandTypes() (int, bool) {
	fmt.Println("\nChoose hand type to practice:")
	fmt.Println("1. Hard totals (no ace or ace = 1)")
	fmt.Println("2. Soft totals (ace = 11)")
	fmt.Println("3. Pairs")
	fmt.Println("0. Cancel")
	fmt.Print("\nChoice (0-3): ")

	reader := bufio.NewReader(os.Stdin)
	input, err := reader.ReadString('\n')
	if err != nil {
		return 0, false
	}

	choice, err := strconv.Atoi(strings.TrimSpace(input))
	if err != nil {
		return 0, false
	}

	if choice == 0 {
		return 0, false
	}

	if choice < 1 || choice > 3 {
		return 0, false
	}

	return choice, true
}
